### PR TITLE
Correctly update bindings of decorated class declarations

### LINKED ISSUE #8559: `Duplicate declaration` error in 7.0.0 with decorators, nested functions, ES6 modules, and a node target
## Bug Report

**Current Behavior**
Transpilation of the code fails with the following error.

```
TypeError: .../babel-duplicate-definition-error/index.js: Duplicate declaration "Foo"
  3 | export default function wrap() {
  4 |   return function() {
> 5 |           class Foo {
    |                 ^
  6 |                   @autobind
  7 |                   method() {}
  8 |           }
    at File.buildCodeFrameError (/private/tmp/babel-duplicate-definition-error/node_modules/@babel/core/lib/transformation/file/file.js:259:12)
    at Scope.checkBlockScopedCollisions (/private/tmp/babel-duplicate-definition-error/node_modules/@babel/traverse/lib/scope/index.js:347:22)
    at Scope.registerBinding (/private/tmp/babel-duplicate-definition-error/node_modules/@babel/traverse/lib/scope/index.js:504:16)
    at Scope.registerDeclaration (/private/tmp/babel-duplicate-definition-error/node_modules/@babel/traverse/lib/scope/index.js:444:14)
    at Object.BlockScoped (/private/tmp/babel-duplicate-definition-error/node_modules/@babel/traverse/lib/scope/index.js:189:28)
    at Object.newFn (/private/tmp/babel-duplicate-definition-error/node_modules/@babel/traverse/lib/visitors.js:230:17)
    at NodePath._call (/private/tmp/babel-duplicate-definition-error/node_modules/@babel/traverse/lib/path/context.js:53:20)
    at NodePath.call (/private/tmp/babel-duplicate-definition-error/node_modules/@babel/traverse/lib/path/context.js:36:14)
    at NodePath.visit (/private/tmp/babel-duplicate-definition-error/node_modules/@babel/traverse/lib/path/context.js:88:12)
    at TraversalContext.visitQueue (/private/tmp/babel-duplicate-definition-error/node_modules/@babel/traverse/lib/context.js:118:16)
```

**Input Code**
Executable gist: https://gist.github.com/dantman/81ab0e3f823ce287d1bd8b2a65b4ad52

* Must use `@babel/preset-env` with a target like `node: 8`
* Must use `@babel/plugin-proposal-decorators` with `legacy: true`

```js
import {autobind} from 'core-decorators';

export default function wrap() {
	return function() {
		class Foo {
			@autobind
			method() {}
		}

		return Foo;
	};
}
```

**Expected behavior/code**
A clear and concise description of what you expected to happen (or code).

**Babel Configuration (.babelrc, package.json, cli command)**

```json
{
  "presets": [
    [
      "@babel/preset-env",
      {
        "targets": {
          "node": 8
        }
      }
    ]
  ],
  "plugins": [
    ["@babel/plugin-proposal-decorators", {"legacy": true}]
  ]	
}
```
```json
{
  "scripts": {
    "start": "babel index.js"
  },
  "devDependencies": {
    "@babel/cli": "^7.0.0",
    "@babel/core": "^7.0.0",
    "@babel/plugin-proposal-decorators": "^7.0.0",
    "@babel/preset-env": "^7.0.0"
  },
  "dependencies": {
    "core-decorators": "^0.20.0"
  }
}
```
```shell
npm install
npm start # babel index.js
```

**Environment**
- Babel version(s): v7.0.0
- Node/npm version: Tested and confirmed in Node 8/9/10 with npm 5.6.0
- OS: OSX 10.12.16
- Monorepo: no
- How you are using Babel: `cli` and `jest`

**Additional context**
#8525 may be related, however I'm reporting it separately because this error happens with a different type of code and does not have "(This is an error on an internal node. Probably an internal error.)" in the error message.

If you make any of the following changes the error goes away:

* Remove the `@autobind` decorator
* Remove the `export default`
* Remove the `return function() {}` so that `wrap` returns `Foo` directly.
* Remove the `node` target from `env` (i.e. change the target so `env` does not emit a number of transforms)

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #8559 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

> This commit adds a small utility (Scope#updateOwnBinding) which
> should be used whenever you replace a declaration with a new
> delaration which defines the same binding (e.g. `class Foo {}` ->
> `let Foo = class Foo {}`). Since our scope system isn't currently
> synchrinized with the AST, doing this automatically would add too
> much overhead to the `replaceWith*` functions.
>
> That function probably needs to be used in many more places, but we
> can add them when users find bugs related to non-updated bingins.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/8566"></a>

### CODE REVIEW

**Comment by jsg2021** [comment]
Does this fix the case where some class fields get transformed not respecting "loose: true" with decorators (legacy)?

I've been trying to create a minimum repro... but, hopefully this it?

**Comment by jsg2021** [comment]
Oh, looking at the issue mentioned in the OP... this isn't my issue. :(

**Comment by TrejGun** [comment]
any chance to get this merged?

**Comment by agrasley** [comment]
I'm running into this issue while trying to upgrade to Babel 7. Would appreciate an update.

**Comment by VincentMolinie** [comment]
Can I help in some way on that subject ? Could be nice to see this merge, it has been 4 years now 😅

**Review by liuxingbaoyu — APPROVED** [review]

File: packages/babel-plugin-proposal-decorators/test/fixtures/legacy-regression/8559/options.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    ["proposal-decorators", { "legacy": true }]
+  ],
+  "presets": [
+    ["env", {
+      "targets": {
+        "node": 8
+      }
+    }]
+  ]
+}
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-decorators/test/fixtures/legacy-regression/8559/options.json:12]:
```suggestion
{
  "targets": { "node": 8 },
  "plugins": [
    ["proposal-decorators", { "version": "legacy" }]
  ],
  "presets": ["env"]
}
```

**Pull request merged**